### PR TITLE
Fixed #26472 -- Added silence_checks().

### PR DESCRIPTION
**Pull request by adamchainz**
ticket-26472

### CODE REVIEW

File: django/core/checks/messages.py [hunk 1]
```diff
@@ -74,3 +87,16 @@ def __init__(self, *args, **kwargs):
 class Critical(CheckMessage):
     def __init__(self, *args, **kwargs):
         super().__init__(CRITICAL, *args, **kwargs)
+
+
+silenced_object_checks = defaultdict(WeakValueDictionary)
```
Comment (adamchainz) [inline django/core/checks/messages.py:92]: Using a weak dictionary so that silencing a check for a temporary object (e.g. a temporary model or field in a test) doesn't cause it to have a permanent reference that keeps it sticking around forever.

Using a `defaultdict(WeakValueDictionary)` rather than `WeakKeyDictinoary` in order to support non-hashable objects (or objects whose hash *changes*... I found out that model `Field` does this, which is naughty, will file a ticket).

There's a proposal to add such a version of `WeakKeyDictionary` that similarly uses `id(obj)`, but no movement: https://bugs.python.org/issue44140

File: docs/releases/4.1.txt [hunk 2]
```diff
@@ -253,6 +253,12 @@ Signals
   :data:`~django.db.models.signals.post_delete` signals now dispatch the
   ``origin`` of the deletion.
 
+System Checks
+~~~~~~~~~~~~~
```
Comment (adamchainz) [inline docs/releases/4.1.txt:257]: I added a section because I didn't think any others were appropriate.

File: django/core/checks/messages.py [hunk 3]
```diff
@@ -48,7 +52,16 @@ def is_serious(self, level=ERROR):
 
     def is_silenced(self):
         from django.conf import settings
-        return self.id in settings.SILENCED_SYSTEM_CHECKS
+
+        if self.id is None:
+            return False
```
Comment (adamchainz) [inline django/core/checks/messages.py:57]: Fixing an edge case where `SILENCED_SYSTEM_CHECKS = [None]` would silence any number of checks that don't have id's.

**Review by adamchainz — COMMENTED** [review]
The third commit only tidies up Django's test suite - you may want to ignore it for initial review.

**Comment by charettes** [comment]
Thanks for this!

This approach seems promising for granularly silencing checks related to fields but I'm curious how it would work for [models](https://docs.djangoproject.com/en/4.0/ref/checks/#models), [urls](https://docs.djangoproject.com/en/4.0/ref/checks/#urls), and [admin](https://docs.djangoproject.com/en/4.0/ref/checks/#admin) and other checks. We should at least have test coverage for that.

I think that's where the _settings_ approach was shining in that it allowed for entries like `('urls.W002', 'some-pattern')` or `('fields.E006', 'model.field')  # this is a false positive` in `SILENCED_SYSTEM_CHECKS` without Python gymnastics. In the end system checks are very much a project things and I fear that this approach might encourage third-party apps that ship code causing check violations to use this pattern instead of overriding methods such as `Field.check`. Is this something you had in mind?

File: tests/runtests.py [hunk 4]
```diff
@@ -232,9 +232,6 @@ def setup_collect_tests(start_at, start_after, test_labels=None):
     # tests.
     log_config['loggers']['django']['level'] = 'ERROR'
     settings.LOGGING = log_config
-    settings.SILENCED_SYSTEM_CHECKS = [
-        'fields.W342',  # ForeignKey(unique=True) -> OneToOneField
-    ]
```
Comment (carltongibson) [inline tests/runtests.py:237]: Can we drop this change, using `silence_checks` with `override_settings` just on specific tests? (We're not removing the current approach no?) — It leads to a lot of noise on the PR. 🤔

**Review by carltongibson — COMMENTED** [review]
Hey @adamchainz — are you up for working on this currently? Did you consider Simon's questions? 

I do worry this is kind of backwards for the use-case reported on the ticket: Upgrading a project that needs many fields updated; only option was to silence the whole check but the desire was to **opt-in** for new fields. Here we have an **opt-out**, which would still require editing every existing model/field/url/... in the whole project. (I start to think about responses to that, and then wonder about the price of admission. … 🤔) 

I'm going to mark this as PNI. If you want to rebase and uncheck when you're ready, I'll take another look. Thanks.
